imports.py

```diff
@@ -12,193 +10,283 @@
+        #return x.replace('/', f'{delim}slash')
+        return x.replace('/', '*slash')
+
+    @staticmethod
+    def decode_filename_str(x):
+        """Convert from friendly system filename"""
+        #return x.replace(f'{delim}slash', '/')
+        return x.replace('*slash', '/')
+
+    def url_and_params_to_filename(url: str, params: Dict) -> str:
+        """Convert to filename"""
+        x = url
+        if params:
+            x += '___params_'
+            for k, v in params.items():
+                x += '__' + k + '_' + v
+        filename = CacheFhir.encode_filename_str(x) + '.json'
+        return filename
+    
+    def cache_write(url: str, params: Dict, response: Dict):
+        """Write to cache"""
+        filename = CacheFhir.url_and_params_to_filename(url, params)
+        if not os.path.exists(CACHE_DIR):
+            os.makedirs(CACHE_DIR)
+        path = os.path.join(CACHE_DIR, filename)
+        with open(path, 'w') as f:
+            json.dump(response, f, indent=4)
+
+    def cache_read(url: str, params: Dict) -> Union[Dict, None]:
+        """Read from cache"""
+        filename = CacheFhir.url_and_params_to_filename(url, params)
+        path = os.path.join(CACHE_DIR, filename)
+        if os.path.exists(path):
+            with open(path, 'r') as f:
+                data: Dict = json.load(f)
+                return data
+        return None
+
 
 
+
+class TimsClient:
+    """TIMS Client"""
+
+    def __init__(self, base_url: str, use_cache=False, verbose=False):
+        self.base_url = base_url
+        self.use_cache = use_cache
+        self.verbose=verbose
+        self.http_client = httpx.Client()
+        
+    def is_response_ok(self, response: Response) -> bool:
+        """Decode status"""
+        if response is None:
+            print("NO RESPONSE")
+            return False
+        else:
+            if  200 <= response.status_code < 300:
+                if self.verbose:
+                    print(f"HTTP STATUS OK {response.status_code}")
+                return True
+            else:
+                print(f"error: {response.status_code}")
+                return False
+
+    def do_hapi_get_request(self, internal_path: str, in_params: Dict = None) -> Dict:
+        """GET request"""
+        url = self.base_url + internal_path
+        response_dict = None
+        if self.use_cache:
+            print(" * USING CACHE * ")
+            response_dict = CacheFhir.cache_read(url, in_params)
+        if response_dict:    
+            return response_dict
+        else:
+            if self.use_cache and self.verbose:
+                print("Cache empty for this request, fetching from server")
+            if self.verbose:
+                print("GETTING")
+            rq = httpx.Request("GET", url, params=in_params, headers=REQUEST_HEADERS)
+            if self.verbose:
+                print(rq.url)
+            response: Response = None
+            try:
+                if self.verbose:
```

Comment:
The server was being fussy and I got an exception here, so I added the catch. The exception type might be too wide...

Reply:
yeah, too wide for now, but that's ok.

i was surprised that so many of my ad hoc attempts failed when trying to demonstrate the notebook live today. im guessing that in a lot of cases it wasn't that I got an error back, but that the JSON coming back was a different shape than expected. not sure

in any case, I don't aim to make the code behind the notebook into a super robust client. I'm pretty happy with HAPI's swagger UI for that